File: algorithms/algorithms-cplusplus/leetcode/Question_0090_Subsets_II.cc
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
 public:
  vector<vector<int>> subsetsWithDup(vector<int>& nums) {
    if (nums.empty()) {
      return vector<vector<int>>();
    }
    vector<vector<int>> res(1);
    sort(nums.begin(), nums.end());
    int pre = nums[0], size = 1;
    for (int i = 0; i < nums.size(); i++) {
      if (pre != nums[i]) {
        pre = nums[i];
        size = res.size();
      }
      int newSize = res.size();
      for (int j = newSize - size; j < newSize; j++) {
        vector<int> tmp = res[j];
        tmp.push_back(nums[i]);
        res.push_back(tmp);
      }
    }
    return res;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0090_Subsets_II.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Question_0090_Subsets_II {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        if (nums == null || nums.length == 0) {
            return res;
        }
        Arrays.sort(nums);
        res.add(new ArrayList<>());
        int size = 1;
        for (int i = 0; i < nums.length; i++) {
            if (i > 0 && nums[i] != nums[i - 1]) {
                size = res.size();
            }
            int newSize = res.size();
            for (int j = newSize - size; j < newSize; j++) {
                List<Integer> tmp = new ArrayList<>(res.get(j));
                tmp.add(nums[i]);
                res.add(tmp);
            }
        }
        return res;
    }
}
